### PR TITLE
HTTP proxy support for k3s/kim

### PR DESCRIPTION
**Pull request by mook-as**
This adds HTTP proxy support, so that kubernetes (and kim) pulls images via the proxy.  This depends on #403 to land first.

### CODE REVIEW

**Review by mattfarina — CHANGES_REQUESTED** [review]
I have not figured out what's causing it but I have 2 problems...

- When RD starts the prefs window does not open
- Kubernetes never starts and no dialog opens to provide me with more information

The only issue I found was in the background.log file...

```
UnhandledRejectionWarning: [Error: ENOENT: no such file or directory, open '/Users/mfarina/code/rancher-sandbox/rd/dev-app-update.yml'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/mfarina/code/rancher-sandbox/rd/dev-app-update.yml'
}
```

I get this message on main where I do not experience the same error. The UI does not present an error.

**Comment by mattfarina** [comment]
To add some more detail, I tested with the CI build for this PR. Everything loaded fine but I was unable to pull the builder image for kim...

My mitm proxy configured for RD (and applied with a reset)...
Screen Shot 2021-08-04 at 3 10 07 PM

My error from `kubectl describe`...

```
  Warning  Failed     26s (x2 over 39s)  kubelet            Failed to pull image "docker.io/moby/buildkit:v0.8.3": rpc error: code = Unknown desc = failed to pull and unpack image "docker.io/moby/buildkit:v0.8.3": failed to resolve reference "docker.io/moby/buildkit:v0.8.3": failed to do request: Head "https://registry-1.docker.io/v2/moby/buildkit/manifests/v0.8.3": proxyconnect tcp: dial tcp [::1]:8080: connect: connection refused
  Warning  Failed     26s (x2 over 39s)  kubelet            Error: ErrImagePull
```

File: src/pages/K8s.vue [hunk 1]
```diff
@@ -24,6 +24,32 @@
...
+        placeholder="http://proxy.example.com:8080"
+        @blur="handleUpdateProxy('httpsProxy')"
+      />
+      <labeled-input
+        :value="settings.kubernetes.noProxy"
+        label="No proxy for"
+        type="string"
+        @input="handleUpdateNoProxy"
+      />
+    </div>
```
[the remark this answers is not in the document]
Comment (mook-as) [reply src/pages/K8s.vue:52]: Should be fixed in 8a221cf70ae0bc8f4a3601da374961218f7df28d.

**Comment by mook-as** [comment]
Using `localhost` as the proxy will not work, since that's from the perspective of the VM… and therefore it will try to look for a proxy there, instead of on your host.  You'll need to use the host's IP address instead.

**Comment by mook-as** [comment]
@mattfarina brings up the good point that we have no way of adding CA certificates yet, so attempting to set up a proxy _will_ break the user, since the interesting parts that this will resolve are pulling from https servers (for images).  So this is blocked by #305 for now.

**Comment by jandubois** [comment]
> we have no way of adding CA certificates yet

If you are just looking to set this up for testing, then you can do this by editing the `lima.yaml` manually and add something like this (merge with existing `mounts` and `provision` lists):

```yaml
mounts:
- location: /dir/with/certs
  writable: false

provision:
- mode: system
  script: |
    #!/bin/sh
    mkdir -p /usr/local/share/ca-certificates
    cp /dir/with/certs/company.crt /usr/local/share/ca-certificates/
    update-ca-certificates
```

You will need to stop and restart the VM for the changes to take effect.

**Comment by mook-as** [comment]
Oh, I have been adding CAs for testing by patching the code to do `limactl shell env http_proxy=… wget -O … http://mitm.it/cert.pem` + `limactl shell update-ca-certificates` just fine, but that's not a reasonable thing to get users to do :)

**Comment by ericpromislow** [comment]
After I change an http or https proxy field, shouldn't I see a notification to reset kubernetes? Or does this take effect immediately (ignoring changes involving certs).

**Comment by mook-as** [comment]
Rebased to resolve conflicts (though GitHub doesn't seem happy with it).

I _do_ get a notification:
Notification when http port has changed, requesting a restart

I had to rebase to get this, though, as for some reason the VM wasn't starting correctly for me (and the restart message never shows up if the VM isn't correctly running).

**Comment by ericpromislow** [comment]
Breakage on the reset notification: https://github.com/rancher-sandbox/rancher-desktop/issues/468

You need to modify the `cmp` method in lima.ts so it does `==` when the types differ (or maybe there's a JSer way to do that). Or figure out why sometimes the port is a string and sometimes it's a number.

**Comment by mook-as** [comment]
With some distance, I'm starting to feel that this isn't going to be the correct way to go (since we'll need to deal with things like [proxy authentication](https://github.com/rancher-sandbox/rancher-desktop/issues/583)).

I think we may be better going down the route we're planning to go with DNS on lima:
- A checkbox to enable proxy support (default off)
- If it's checked, set up a local (non-caching) proxy, that just preserves all HTTP headers and requests things via the platform-native stack.  This implies `CGO` everywhere.

**Comment by jandubois** [comment]
Sorry, I've not been paying attention to this issue lately, but was planning to check in already, as I've recently added https://github.com/lima-vm/lima/pull/266 to Lima: all system proxy settings are forwarded automatically to the guest instance as environment variables via `/etc/environment`.

Settings are *not* copied when they have a user name, as the password will be stored in a keychain. You can encode username and password in the proxy URL, if you are willing to expose them in the environment. This works even via system settings.

I know this doesn't cover everything, but it is pretty much what a non-browser app would have access to on the host.

Maybe we should have a brainstorming session on what else is needed. Aren't the certs for MitM proxies already copied?

**Comment by dragosh** [comment]
Hi @jandubois, Will the automatic system proxy settings work if, for instance, you have an auth proxy server, like [Preproxy](https://hamstergene.github.io/macapps/preproxy/) or [cntlm](http://cntlm.sourceforge.net/) wherein that case you will have `http(s)_proxy` set to: `http://localhost:3128` (host) for example and then this will be forwarded to VM as `http://localhost:3128` and in that case will not work.

👉 `startup log`
```
...
INFO[0003] [hostagent] time="2021-10-19T10:13:52+02:00" level=info msg="Overriding \"https_proxy\" value \"http://192.168.5.2:3128\" with \"http://localhost:3128\" from limactl process environment"
....
```
👉 `serial.log`
```
42.991220] cloud-init[1314]: W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/hirsute-updates/InRelease  Unable to connect to localhost:3128:
```

PS: prior to `v0.7.x` I use to have that explicit in the lima.yml config file and used to work:

```yml
env:
  http_proxy: "http://192.168.5.2:3128" <- host IP
  https_proxy: "http://192.168.5.2:3128"
```

**Is there a way to give prio. order first to the config and then to the system (host) proxy for this type of proxy authentication?**

**Comment by mook-as** [comment]
Closing this — we're going to go with a different approach, I believe.

**Pull request closed without merging**